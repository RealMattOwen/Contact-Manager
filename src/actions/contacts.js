// ADD_CONTACT
export const addContact = contactDetails => ({
    type: 'ADD_CONTACT',
    contactDetails
});

export const startAddContact = contactDetails => {
    return dispatch => {
        return new Promise((resolve, reject) => {
            resolve(dispatch(addContact(contactDetails)));
        });
    };
};

// EDIT_CONTACT
export const editContact = (id, updates) => ({
    type: 'EDIT_CONTACT',
    id,
    updates
});

export const startEditContact = (id, updates) => {
    return dispatch => {
        return new Promise((resolve, reject) => {
            resolve(dispatch(editContact(id, updates)));
        });
    };
};

// REMOVE_CONTACT
export const removeContact = ({id} = {}) => ({
    type: 'REMOVE_CONTACT',
    id
});

export const startRemoveContact = ({id} = {}) => {
    return dispatch => {
        return new Promise((resolve, reject) => {
            resolve(dispatch(removeContact({ id })));
        });
    };
};

// SET_CONTACTS
export const setContacts = contacts => ({
    type: 'SET_CONTACTS',
    contacts
});