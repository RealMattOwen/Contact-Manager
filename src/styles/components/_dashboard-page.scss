.list-item {
	display: block;
	text-decoration: none;
	transition: background .3s ease;

	&:hover, &:focus {
		background-color: darken(white, 7%);
	}
}