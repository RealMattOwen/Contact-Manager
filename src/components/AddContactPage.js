import React from 'react';
import { connect } from 'react-redux';
import { history } from '../routers/AppRouter';
import AppBar from 'material-ui/AppBar';
import BackIcon from 'material-ui/svg-icons/navigation/arrow-back';
import DoneIcon from 'material-ui/svg-icons/action/done';
import IconButton from 'material-ui/IconButton';
import Paper from 'material-ui/Paper';
import TextField from 'material-ui/TextField';
import { startAddContact } from '../actions/contacts';

export class AddContactPage extends React.Component {
    state = this.props.contacts.length !== 0 ? {
        age: undefined,
        firstName: undefined,
        id: undefined,
        initials: undefined,
        lastName: undefined
    } : {};
    componentWillMount() {
        const contacts = JSON.parse(localStorage.getItem('contacts'));

        if (contacts) {
            contacts.length !== 0 && this.props.contacts.length === 0 ? history.push('/') : this.setState(() => ({id: this.props.contacts.length + 1}));
        } else {
            this.setState(() => ({id: 1}));
        }
    }
    handleClick = () => {
        this.setState(() => {
            return {
                initials: this.state.firstName.charAt(0).concat(this.state.lastName.charAt(0))
            }
        }, () => {
            this.props.startAddContact(this.state)
                .then(() => {
                    localStorage.setItem('contacts', JSON.stringify(this.props.contacts));
                    history.push('/');
                });
        });
    };
    render() {
        return (
            <div>
                <AppBar
                    iconElementLeft={<IconButton onClick={() => history.push('/')}><BackIcon /></IconButton>}
                    iconElementRight={<IconButton type="button" onClick={this.handleClick}><DoneIcon /></IconButton>}
                    title="Add new contact"
                />
                <Paper
                    rounded={false}
                    style={{
                        display: 'inline-block',
                        textAlign: 'center'
                    }}
                    zDepth={1}>
                    <form id="contact-details">
                        <TextField
                            hintText="First name"
                            onChange={(e, newValue) => {
                                this.setState(() => {
                                    return {
                                        firstName: newValue
                                    }
                                });
                            }}
                        />
                        <TextField
                            hintText="Last name"
                            onChange={(e, newValue) => {
                                this.setState(() => {
                                    return {
                                        lastName: newValue
                                    }
                                });
                            }}
                        />
                        <TextField
                            hintText="Age"
                            onChange={(e, newValue) => {
                                this.setState(() => {
                                    return {
                                        age: parseInt(newValue)
                                    }
                                });
                            }}
                        />
                    </form>
                </Paper>
            </div>
        )
    }
}

const mapStateToProps = state => ({
    contacts: state.contacts
});

const mapDispatchToProps = dispatch => ({
    startAddContact: contactDetails => dispatch(startAddContact(contactDetails))
});

export default connect(mapStateToProps, mapDispatchToProps)(AddContactPage);