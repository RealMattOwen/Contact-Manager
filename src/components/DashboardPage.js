import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import AddPersonIcon from 'material-ui/svg-icons/social/person-add';
import AppBar from 'material-ui/AppBar';
import Avatar from 'material-ui/Avatar';
import Divider from 'material-ui/Divider';
import FloatingActionButton from 'material-ui/FloatingActionButton';
import List from 'material-ui/List/List';
import ListItem from 'material-ui/List/ListItem';
import { setContacts } from '../actions/contacts';

export class DashboardPage extends React.Component {
    componentDidMount() {
        const contacts = JSON.parse(localStorage.getItem('contacts'));

        if (contacts) {
            this.props.setContacts(contacts);
        }
    };
    render() {
        return (
            <div>
                <AppBar
                    showMenuIconButton={false}
                    title="Contacts"
                />
                <List>
                    {this.props.contacts.length !== 0 ? this.props.contacts.map(({age, firstName, id, initials, lastName}) => (
                        <div key={id}>
                            <Link className="list-item" to={`/edit/${id}`}>
                                <ListItem
                                    disabled={true}
                                    leftAvatar={<Avatar style={{fontSize: 18}}>{initials}</Avatar>}
                                >
                                    {firstName} {lastName}, {age}
                                </ListItem>
                            </Link>
                            <Divider />
                        </div>
                    )) : (
                        <div>
                            <ListItem disabled={true}>No contacts</ListItem>
                        </div>
                    )}
                </List>
                <Link to={"/add"}>
                    <FloatingActionButton
                        style={{
                            margin: 0,
                            top: 'auto',
                            right: 20,
                            bottom: 20,
                            left: 'auto',
                            position: 'fixed'
                        }}>
                        <AddPersonIcon />
                    </FloatingActionButton>
                </Link>
            </div>
        );
    };
}

const mapStateToProps = state => ({
    contacts: state.contacts
});

const mapDispatchToProps = dispatch => ({
    setContacts: contacts => dispatch(setContacts(contacts))
});

export default connect(mapStateToProps, mapDispatchToProps)(DashboardPage);