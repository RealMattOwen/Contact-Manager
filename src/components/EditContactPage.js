import React from 'react';
import { connect } from 'react-redux';
import { history } from '../routers/AppRouter';
import AppBar from 'material-ui/AppBar';
import BackIcon from 'material-ui/svg-icons/navigation/arrow-back';
import DoneIcon from 'material-ui/svg-icons/action/done';
import IconButton from 'material-ui/IconButton';
import Paper from 'material-ui/Paper';
import RaisedButton from 'material-ui/RaisedButton';
import TextField from 'material-ui/TextField';
import { startEditContact, startRemoveContact } from '../actions/contacts';

export class EditContactPage extends React.Component {
    state = this.props.contacts.length !== 0 ? {
        age: this.props.contact.age,
        firstName: this.props.contact.firstName,
        initials: this.props.contact.initials,
        lastName: this.props.contact.lastName
    } : {};
    componentWillMount() {
        this.props.contacts.length === 0 ? history.push('/') : undefined;
    }
    handleClick = () => {
        this.setState(() => {
            return {
                initials: this.state.firstName.charAt(0).concat(this.state.lastName.charAt(0))
            }
        }, () => {
            this.props.startEditContact(this.props.contact.id, { ...this.state })
                .then(() => {
                    localStorage.setItem('contacts', JSON.stringify(this.props.contacts));
                    history.push('/');
                });
        });
    };
    onRemove = () => {
        this.props.startRemoveContact({ id: this.props.contact.id })
            .then(() => {
                localStorage.setItem('contacts', JSON.stringify(this.props.contacts));
                history.push('/');
            });
    };
    render() {
        return (
            <div>
                <AppBar
                    iconElementLeft={<IconButton onClick={() => history.push('/')}><BackIcon /></IconButton>}
                    iconElementRight={<IconButton onClick={this.handleClick}><DoneIcon /></IconButton>}
                    title="Edit contact"
                />
                <Paper
                    rounded={false}
                    style={{
                        textAlign: 'center',
                        display: 'inline-block',
                    }}
                    zDepth={1}>
                    <form id="contact-details">
                        <TextField
                            hintText="First name"
                            onChange={(e, newValue) => {
                                this.setState(() => {
                                    return {
                                        firstName: newValue
                                    }
                                });
                            }}
                            value={this.state.firstName}
                        />
                        <TextField
                            hintText="Last name"
                            onChange={(e, newValue) => {
                                this.setState(() => {
                                    return {
                                        lastName: newValue
                                    }
                                });
                            }}
                            value={this.state.lastName}
                        />
                        <TextField
                            hintText="Age"
                            onChange={(e, newValue) => {
                                this.setState(() => {
                                    return {
                                        age: parseInt(newValue)
                                    }
                                });
                            }}
                            value={this.state.age}
                        />
                        <RaisedButton label="Remove" onClick={this.onRemove} style={{margin: 12}} />
                    </form>
                </Paper>
            </div>
        )
    }
}

const mapStateToProps = (state, props) => ({
    contact: state.contacts.find(contact => contact.id === parseInt(props.match.params.id)),
    contacts: state.contacts
});

const mapDispatchToProps = dispatch => ({
    startEditContact: (id, updates) => dispatch(startEditContact(id, updates)),
    startRemoveContact: id => dispatch(startRemoveContact(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditContactPage);