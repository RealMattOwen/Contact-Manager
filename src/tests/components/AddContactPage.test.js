import React from 'react';
import { shallow } from 'enzyme';
import { AddContactPage } from '../../components/AddContactPage';
import contacts from '../fixtures/contacts';

test('should render AddContactPage correctly', () => {
    const wrapper = shallow(<AddContactPage contacts={contacts} />);
    expect(wrapper).toMatchSnapshot();
});