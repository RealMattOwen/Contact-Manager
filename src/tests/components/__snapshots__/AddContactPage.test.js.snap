// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddContactPage correctly 1`] = `
<div>
  <AppBar
    iconElementLeft={
      <IconButton
        disableTouchRipple={false}
        disabled={false}
        iconStyle={Object {}}
        onClick={[Function]}
        tooltipPosition="bottom-center"
        touch={false}
      >
        <NavigationArrowBack />
      </IconButton>
    }
    iconElementRight={
      <IconButton
        disableTouchRipple={false}
        disabled={false}
        iconStyle={Object {}}
        onClick={[Function]}
        tooltipPosition="bottom-center"
        touch={false}
        type="button"
      >
        <ActionDone />
      </IconButton>
    }
    showMenuIconButton={true}
    title="Add new contact"
    zDepth={1}
  />
  <Paper
    circle={false}
    rounded={false}
    style={
      Object {
        "display": "inline-block",
        "textAlign": "center",
      }
    }
    transitionEnabled={true}
    zDepth={1}
  >
    <form
      id="contact-details"
    >
      <TextField
        disabled={false}
        floatingLabelFixed={false}
        fullWidth={false}
        hintText="First name"
        multiLine={false}
        onChange={[Function]}
        rows={1}
        type="text"
        underlineShow={true}
      />
      <TextField
        disabled={false}
        floatingLabelFixed={false}
        fullWidth={false}
        hintText="Last name"
        multiLine={false}
        onChange={[Function]}
        rows={1}
        type="text"
        underlineShow={true}
      />
      <TextField
        disabled={false}
        floatingLabelFixed={false}
        fullWidth={false}
        hintText="Age"
        multiLine={false}
        onChange={[Function]}
        rows={1}
        type="text"
        underlineShow={true}
      />
    </form>
  </Paper>
</div>
`;
